india_forecast_app/adjuster.py

```diff
@@ -0,0 +1,161 @@
+    start_utc_minute_rounded = (
+        cast(func.date_part("minute", GenerationSQL.start_utc), INT)
+        / 15
+        * text("interval '15 min'")
+    )
+    start_utc_hour = func.date_trunc("hour", GenerationSQL.start_utc)
+    generation_start_utc = start_utc_hour + start_utc_minute_rounded
+    query = query.filter(generation_start_utc == ForecastValueSQL.start_utc)
+
+    # only include the last x days
+    query = query.filter(ForecastValueSQL.start_utc >= start_datetime)
+    query = query.filter(GenerationSQL.start_utc >= start_datetime)
+
+    # filter on site
+    query = query.filter(ForecastSQL.site_uuid == site_uuid)
+    query = query.filter(GenerationSQL.site_uuid == site_uuid)
+
+    # filter on created_utc
+    query = query.filter((func.extract("hour", ForecastSQL.created_utc) == hour))
+
+    if ml_model_name is not None:
+        query = query.join(MLModelSQL)
+        query = query.filter(MLModelSQL.name == ml_model_name)
+
+    # group by forecast horizon
+    query = query.group_by(ForecastValueSQL.horizon_minutes)
+
+    # order by forecast horizon
+    query = query.order_by(ForecastValueSQL.horizon_minutes)
+
+    me = query.all()
+
+    me_df = pd.DataFrame(me, columns=["me_kw", "horizon_minutes"])
+
+    return me_df
+
+
+def adjust_forecast_with_adjuster(
+    db_session, forecast_meta: dict, forecast_values_df: pd.DataFrame, ml_model_name: str
+):
+    """
+    Adjust forecast values with ME values
+
+    Args:
+    db_session: sqlalchemy session
+    forecast_meta: forecast metadata
+    forecast_values_df: forecast values dataframe
+    ml_model_name: the ml model name
+
+    """
+    # get the ME values
+    me_values = get_me_values(
+        db_session,
+        forecast_meta["timestamp_utc"].hour,
+        site_uuid=forecast_meta["site_uuid"],
+        ml_model_name=ml_model_name,
+    )
+    log.debug(f"ME values: {me_values}")
+
+    # clip me values by 10% of the capacity
+    site = get_site_by_uuid(db_session, forecast_meta["site_uuid"])
+    capacity = site.capacity_kw
+    me_kw_limit = 0.1 * capacity
+    n_values_above_limit = (me_values["me_kw"] > me_kw_limit).sum()
+    n_values_below_limit = (me_values["me_kw"] > me_kw_limit).sum()
+    me_values["me_kw"].clip(lower=-0.1 * capacity, upper=0.1 * capacity, inplace=True)
+    log.debug(
+        f"ME values clipped: There were {n_values_above_limit} values above the limit and "
+        f"{n_values_below_limit} values below the limit."
+    )
+
+    # join forecast_values_df with me_values on horizon_minutes
+    forecast_values_df_adjust = forecast_values_df.copy()
+    forecast_values_df_adjust = forecast_values_df_adjust.merge(
+        me_values, on="horizon_minutes", how="left"
+    )
+
+    # if me_kw is null, set to 0
+    forecast_values_df_adjust["me_kw"].fillna(0, inplace=True)
+
+    # adjust forecast_power_kw by ME values
+    forecast_values_df_adjust["forecast_power_kw"] = (
```

Comment:
This is currently applied to all forecast horizons, is that okay or do we want to add some logic to restrict this to be applied to just day ahead horizons?

Reply:
Yea, currently not. I was thinking, perhaps to deploy it and see the behaviour on this. 
You would hope our ME on the smaller forecast horizons is smaller due to live data, satellite e.t.c, (which i saw breifly when looking at me), so this extra bit of logic might not be necessary.